Fixed documentation comment ID handling for dynamic and tuples

svick opened the pull request:
Fixes https://github.com/dotnet/roslyn/issues/34839.
Fixes https://github.com/dotnet/roslyn/issues/48495.

Previously, `DocumentationCommentId.CreateDeclarationId` behaved correctly for named and unnamed tuples, but not for `dynamic`. And `DocumentationCommentId.GetSymbolsForDeclarationId` only behaved correctly for unnamed tuples, but not for named tuples or `dynamic`.

This PR fixes all the cases that didn't behave correctly so that the documentation comment ID represents CLR signature of the member (i.e. `System.Object` for `dynamic` and `System.ValueTuple` for tuples).

333fred commented on src/Compilers/Core/Portable/DocumentationCommentId.cs:
```diff
@@ -649,6 +656,9 @@ private bool IsInScope(ITypeParameterSymbol typeParameterSymbol)
 
         private static class Parser
         {
+            private static readonly SymbolEqualityComparer s_symbolEqualityComparer =
```
We should define this as internal on `SymbolEqualityComparer`, rather than just here.

333fred commented on src/Compilers/Core/CodeAnalysisTest/Symbols/DocumentationCommentIdTests.cs:
```diff
@@ -0,0 +1,75 @@
+            string code = @"
+class C
+{
+    (int i, int) DoStuff() => default;
```
Please add a test that covers tuples in parameters as well.

333fred commented:
Done review pass (commit 1)

svick commented:
@333fred I have addressed your comments and also added a test for `nint`.

333fred commented on src/Compilers/Core/Portable/Symbols/SymbolEqualityComparer.cs:
```diff
@@ -16,16 +16,17 @@ public sealed class SymbolEqualityComparer : IEqualityComparer<ISymbol?>
+        public static readonly SymbolEqualityComparer IncludeNullability = new(TypeCompareKind.ConsiderEverything2); //TODO: should this be explicitly *not* compare everything
 
         // Internal only comparisons:
-        internal static readonly SymbolEqualityComparer ConsiderEverything = new SymbolEqualityComparer(TypeCompareKind.ConsiderEverything);
```
Please don't reformat existing code, we prefer to leave things as they are. It helps make `git blame` more useful by not having otherwise-unrelated PRs in the history.

svick replied:
Ok, I have changed it back. Though I don't like much that this means that the code is probably never going to be updated. 🙂

333fred replied:
We prefer to have git blame be useful 😉.

333fred reviewed (commented):
Done review pass (commit 3)

333fred reviewed (approved):
LGTM. @dotnet/roslyn-compiler for a second review.

333fred commented:
@dotnet/roslyn-compiler for a second review of this community PR.

svick commented:
I believe this should also fix the recently opened https://github.com/dotnet/roslyn/issues/48495, so I have added a test for that and modified "fixes" accordingly.

Youssef1313 commented on src/Compilers/Core/CodeAnalysisTest/Symbols/DocumentationCommentIdTests.cs:
```diff
@@ -0,0 +1,124 @@
+
+            var actualDocId = DocumentationCommentId.CreateDeclarationId(symbol);
+
+            string expectedDocId = "M:C.DoStuff(System.ValueTuple{System.Int32,System.Int32})";
```
I think it worth another test with larger tuples (8 elements and 9 elements) to see if the tuple nesting will work as expected.

svick replied:
Sure, I have added a test case for that.

333fred commented:
Thanks @svick!
